### PR TITLE
Add OkHttp-based server

### PR DESCRIPTION
**Pull request by ejona86**
Needs more tests, but it actually passes AbstractTransportTest and AbstractInteropTest with okhttp and netty clients.

I do think I can split out some of the changes to existing code to make it more clear why those are made. So don't look too closely at them right now.

CC @temawi @YifeiZhuang

### CODE REVIEW

File: okhttp/src/main/java/io/grpc/okhttp/OkHttpServerBuilder.java [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+    if (socketFactory == null) {
+      this.socketFactory = ServerSocketFactory.getDefault();
+    } else {
+      this.socketFactory = socketFactory;
+    }
+    return this;
+  }
+
+  /**
+   * Sets a custom keepalive time, the delay time for sending next keepalive ping after a period of
```
Comment (temawi) [inline okhttp/src/main/java/io/grpc/okhttp/OkHttpServerBuilder.java:168]: Also mention that negative values will be rejected?

File: okhttp/src/main/java/io/grpc/okhttp/OkHttpServerBuilder.java [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+   * inactivity. An unreasonably small value might be increased, and {@code Long.MAX_VALUE} nano
+   * seconds or a very large value will disable keepalive.
+   */
+  public OkHttpServerBuilder keepAliveTime(long keepAliveTime, TimeUnit timeUnit) {
+    Preconditions.checkArgument(keepAliveTime > 0L, "keepalive time must be positive");
+    keepAliveTimeNanos = timeUnit.toNanos(keepAliveTime);
+    keepAliveTimeNanos = KeepAliveManager.clampKeepAliveTimeInNanos(keepAliveTimeNanos);
```
Comment (temawi) [inline okhttp/src/main/java/io/grpc/okhttp/OkHttpServerBuilder.java:175]: KeepAliveManager comments imply to me that "keep alive pings" are used by clients to keep a connection going - how does the OkHttpServer use this value?

File: okhttp/src/main/java/io/grpc/okhttp/OkHttpServerBuilder.java [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+  ObjectPool<Executor> transportExecutorPool = DEFAULT_TRANSPORT_EXECUTOR_POOL;
+  ObjectPool<ScheduledExecutorService> scheduledExecutorServicePool =
+      SharedResourcePool.forResource(GrpcUtil.TIMER_SERVICE);
+
+  ServerSocketFactory socketFactory = ServerSocketFactory.getDefault();
+  long keepAliveTimeNanos = KEEPALIVE_TIME_NANOS_DISABLED;
+  long keepAliveTimeoutNanos = DEFAULT_KEEPALIVE_TIMEOUT_NANOS;
+  int flowControlWindow = DEFAULT_FLOW_CONTROL_WINDOW;
+  int maxInboundMetadataSize = GrpcUtil.DEFAULT_MAX_HEADER_LIST_SIZE;
+  int maxInboundMessageSize = GrpcUtil.DEFAULT_MAX_MESSAGE_SIZE;
```
Comment (temawi) [inline okhttp/src/main/java/io/grpc/okhttp/OkHttpServerBuilder.java:118]: These members have setters to populate them, but why don't I see them used anywhere?

File: okhttp/src/main/java/io/grpc/okhttp/OkHttpServerBuilder.java (hunk 2 above)
Comment (ejona86) [reply okhttp/src/main/java/io/grpc/okhttp/OkHttpServerBuilder.java:175]: [gRFC A8](https://github.com/grpc/proposal/blob/master/A8-client-side-keepalive.md) (ignoring Extending for Basic Health Checking) is probably the best introduction to why we do this stuff. The reasoning for client-side is very similar on server-side, which was covered by [gRFC A9](https://github.com/grpc/proposal/blob/master/A9-server-side-conn-mgt.md) (which doesn't say too much more). In short, there's two potential values: 1) trigger activity so the _network_ (routers, NATs, etc) knows the connection is still in use and 2) detect if the connection has been broken by the network.

File: okhttp/src/main/java/io/grpc/okhttp/OkHttpServerBuilder.java (hunk 1 above)
Comment (ejona86) [reply okhttp/src/main/java/io/grpc/okhttp/OkHttpServerBuilder.java:168]: `io.grpc.ServerBuilder` would be the important place to change that. This is old text from NettyServerBuilder. We're going to do a review of it this Thursday. In this PR I still need to merge with https://github.com/grpc/grpc-java/pull/9176 so I can use `@Override` and such here.

File: okhttp/src/main/java/io/grpc/okhttp/OkHttpServerBuilder.java (hunk 3 above)
Comment (ejona86) [reply okhttp/src/main/java/io/grpc/okhttp/OkHttpServerBuilder.java:118]: I could believe some of these are as-of-yet unused and are part of some TODO/FIXME/should-add-a-TODO. But this particular field is used in `OkHttpServerTransport.Config` and then copied to `OkHttpServerStream.TransportState`.

**Comment by ejona86** [comment]
This needs to be rebased for the CI to be happy. I'll coordinate to determine when to do so.

**Review by temawi — APPROVED** [review]

**Pull request merged**